File: app/src/main/java/com/hiddenramblings/tagmo/browser/adapter/FlaskSlotAdapter.java
package com.hiddenramblings.tagmo.browser.adapter;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.hiddenramblings.tagmo.GlideApp;
import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.Amiibo;
import com.hiddenramblings.tagmo.amiibo.FlaskTag;
import com.hiddenramblings.tagmo.browser.BrowserSettings;
import com.hiddenramblings.tagmo.browser.BrowserSettings.BrowserSettingsListener;
import com.hiddenramblings.tagmo.browser.BrowserSettings.VIEW;

import java.util.ArrayList;

public class FlaskSlotAdapter
        extends RecyclerView.Adapter<FlaskSlotAdapter.FlaskViewHolder>
        implements BrowserSettingsListener {

    private final BrowserSettings settings;
    private final OnAmiiboClickListener listener;
    private ArrayList<Amiibo> flaskAmiibo = new ArrayList<>();

    public FlaskSlotAdapter(BrowserSettings settings, OnAmiiboClickListener listener) {
        this.settings = settings;
        this.listener = listener;
    }

    public void setFlaskAmiibo(ArrayList<Amiibo> amiibo) {
        this.flaskAmiibo = amiibo;
    }
    public void addFlaskAmiibo(ArrayList<Amiibo> amiibo) {
        this.flaskAmiibo.addAll(amiibo);
    }

    @Override
    public void onBrowserSettingsChanged(BrowserSettings newBrowserSettings,
                                         BrowserSettings oldBrowserSettings) { }

    @Override
    public int getItemCount() {
        return null != flaskAmiibo ? flaskAmiibo.size() : 0;
    }

    @Override
    public long getItemId(int i) {
        return Long.parseLong(flaskAmiibo.get(i).getFlaskTail());
    }

    public Amiibo getItem(int i) {
        return flaskAmiibo.get(i);
    }

    @Override
    public int getItemViewType(int position) {
        return settings.getAmiiboView();
    }

    @NonNull
    @Override
    public FlaskViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (VIEW.valueOf(viewType)) {
            case COMPACT:
                return new CompactViewHolder(parent, settings, listener);
            case LARGE:
                return new LargeViewHolder(parent, settings, listener);
            case IMAGE:
                return new ImageViewHolder(parent, settings, listener);
            case SIMPLE:
            default:
                return new SimpleViewHolder(parent, settings, listener);
        }
    }

    @Override
    public void onBindViewHolder(final FlaskViewHolder holder, int position) {
        View highlight = holder.itemView.findViewById(R.id.highlight);
        if (TagMo.getPrefs().flaskActiveSlot() == position) {
            highlight.setBackgroundResource(R.drawable.cardview_outline);
        } else {
            highlight.setBackgroundResource(0);
        }
        holder.itemView.setOnClickListener(view -> {
            if (null != holder.listener) {
                holder.listener.onAmiiboClicked(holder.amiibo);
            }
        });
        if (null != holder.imageAmiibo) {
            holder.imageAmiibo.setOnClickListener(view -> {
                if (null != holder.listener) {
                    if (settings.getAmiiboView() == VIEW.IMAGE.getValue())
                        holder.listener.onAmiiboClicked(holder.amiibo);
                    else
                        holder.listener.onAmiiboImageClicked(holder.amiibo);
                }
            });
        }
        holder.bind(getItem(position));
    }

    protected static abstract class FlaskViewHolder extends RecyclerView.ViewHolder {
        private final BrowserSettings settings;
        private final OnAmiiboClickListener listener;

        public final TextView txtError;
        public final TextView txtName;
        public final TextView txtTagId;
        public final TextView txtAmiiboSeries;
        public final TextView txtAmiiboType;
        public final TextView txtGameSeries;
        public final TextView txtPath;
        public final AppCompatImageView imageAmiibo;

        Amiibo amiibo = null;

        CustomTarget<Bitmap> target = new CustomTarget<>() {
            @Override
            public void onLoadStarted(@Nullable Drawable placeholder) {
                imageAmiibo.setImageResource(0);
            }

            @Override
            public void onLoadFailed(@Nullable Drawable errorDrawable) {
                imageAmiibo.setVisibility(View.INVISIBLE);
            }

            @Override
            public void onLoadCleared(@Nullable Drawable placeholder) {
                imageAmiibo.setVisibility(View.VISIBLE);
            }

            @Override
            public void onResourceReady(@NonNull Bitmap resource, Transition transition) {
                imageAmiibo.setImageBitmap(resource);
                imageAmiibo.setVisibility(View.VISIBLE);
            }
        };

        public FlaskViewHolder(
                View itemView, BrowserSettings settings,
                OnAmiiboClickListener listener) {
            super(itemView);

            this.listener = listener;
            this.settings = settings;

            this.txtError = itemView.findViewById(R.id.txtError);
            this.txtName = itemView.findViewById(R.id.txtName);
            this.txtTagId = itemView.findViewById(R.id.txtTagId);
            this.txtAmiiboSeries = itemView.findViewById(R.id.txtAmiiboSeries);
            this.txtAmiiboType = itemView.findViewById(R.id.txtAmiiboType);
            this.txtGameSeries = itemView.findViewById(R.id.txtGameSeries);
            this.txtPath = itemView.findViewById(R.id.txtPath);
            this.imageAmiibo = itemView.findViewById(R.id.imageAmiibo);
        }

        void bind(final Amiibo item) {
            this.amiibo = item;

            String amiiboHexId;
            String amiiboSeries = "";
            String amiiboType = "";
            String gameSeries = "";
            String amiiboImageUrl = null;

            if (amiibo instanceof FlaskTag)  {
                setAmiiboInfoText(txtName, TagMo.getContext().getString(R.string.blank_tag));
            } else {
                setAmiiboInfoText(txtName, amiibo.name);
                amiiboImageUrl = amiibo.getImageUrl();
            }
            if (settings.getAmiiboView() != VIEW.IMAGE.getValue()) {
                this.txtError.setVisibility(View.GONE);
                this.txtPath.setVisibility(View.GONE);
                if (amiibo instanceof FlaskTag) {
                    this.txtTagId.setVisibility(View.GONE);
                    this.txtAmiiboSeries.setVisibility(View.GONE);
                    this.txtAmiiboType.setVisibility(View.GONE);
                    this.txtGameSeries.setVisibility(View.GONE);
                } else {
                    amiiboHexId = Amiibo.idToHex(amiibo.id);
                    if (null != amiibo.getAmiiboSeries())
                        amiiboSeries = amiibo.getAmiiboSeries().name;
                    if (null != amiibo.getAmiiboType())
                        amiiboType = amiibo.getAmiiboType().name;
                    if (null != amiibo.getGameSeries())
                        gameSeries = amiibo.getGameSeries().name;

                    this.txtTagId.setVisibility(View.VISIBLE);
                    this.txtAmiiboSeries.setVisibility(View.VISIBLE);
                    this.txtAmiiboType.setVisibility(View.VISIBLE);
                    this.txtGameSeries.setVisibility(View.VISIBLE);
                    setAmiiboInfoText(txtTagId, amiiboHexId);
                    setAmiiboInfoText(txtAmiiboSeries, amiiboSeries);
                    setAmiiboInfoText(txtAmiiboType, amiiboType);
                    setAmiiboInfoText(txtGameSeries, gameSeries);
                    if (amiiboHexId.endsWith("00000002") && !amiiboHexId.startsWith("00000000")) {
                        txtTagId.setEnabled(false);
                    }
                }
            }

            if (null == amiiboImageUrl) {
                this.imageAmiibo.setImageResource(R.mipmap.ic_launcher_round);
                this.imageAmiibo.setVisibility(View.VISIBLE);
            } else if (null != this.imageAmiibo) {
                GlideApp.with(itemView).clear(target);
                GlideApp.with(itemView).asBitmap().load(amiiboImageUrl).into(target);
            }
        }

        void setAmiiboInfoText(TextView textView, CharSequence text) {
            textView.setVisibility(View.VISIBLE);
            if (text.length() == 0) {
                textView.setText(R.string.unknown);
                textView.setEnabled(false);
            } else {
                textView.setText(text);
                textView.setEnabled(true);
            }
        }
    }

    static class SimpleViewHolder extends FlaskViewHolder {
        public SimpleViewHolder(
                ViewGroup parent, BrowserSettings settings, OnAmiiboClickListener listener
        ) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_simple_card, parent, false),
                    settings, listener
            );
        }
    }

    static class CompactViewHolder extends FlaskViewHolder {
        public CompactViewHolder(
                ViewGroup parent, BrowserSettings settings, OnAmiiboClickListener listener
        ) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_compact_card, parent, false),
                    settings, listener
            );
        }
    }

    static class LargeViewHolder extends FlaskViewHolder {
        public LargeViewHolder(
                ViewGroup parent, BrowserSettings settings, OnAmiiboClickListener listener
        ) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_large_card, parent, false),
                    settings, listener
            );
        }
    }

    static class ImageViewHolder extends FlaskViewHolder {
        public ImageViewHolder(
                ViewGroup parent, BrowserSettings settings, OnAmiiboClickListener listener
        ) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_image_card, parent, false),
                    settings, listener
            );
        }
    }

    public interface OnAmiiboClickListener {
        void onAmiiboClicked(Amiibo amiibo);

        void onAmiiboImageClicked(Amiibo amiibo);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/browser/adapter/FlaskSlotAdapter.kt
package com.hiddenramblings.tagmo.browser.adapter

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.hiddenramblings.tagmo.GlideApp
import com.hiddenramblings.tagmo.Preferences
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.amiibo.Amiibo
import com.hiddenramblings.tagmo.amiibo.AmiiboManager.hasSpoofData
import com.hiddenramblings.tagmo.amiibo.FlaskTag
import com.hiddenramblings.tagmo.browser.BrowserSettings
import com.hiddenramblings.tagmo.browser.BrowserSettings.BrowserSettingsListener
import com.hiddenramblings.tagmo.browser.BrowserSettings.VIEW
import com.hiddenramblings.tagmo.browser.adapter.FlaskSlotAdapter.FlaskViewHolder

class FlaskSlotAdapter(
    private val settings: BrowserSettings, private val listener: OnAmiiboClickListener
) : RecyclerView.Adapter<FlaskViewHolder>(), BrowserSettingsListener {
    var mPrefs = Preferences(TagMo.appContext)
    private var flaskAmiibo: ArrayList<Amiibo?> = arrayListOf()
    fun setFlaskAmiibo(amiibo: ArrayList<Amiibo?>) {
        flaskAmiibo = amiibo
    }

    fun addFlaskAmiibo(amiibo: ArrayList<Amiibo?>) {
        flaskAmiibo.addAll(amiibo)
    }

    override fun onBrowserSettingsChanged(
        newBrowserSettings: BrowserSettings?,
        oldBrowserSettings: BrowserSettings?
    ) {
    }

    override fun getItemCount(): Int {
        return flaskAmiibo.size
    }

    override fun getItemId(i: Int): Long {
        return (flaskAmiibo[i]?.flaskTail ?: "").toLong()
    }

    fun getItem(i: Int): Amiibo? {
        return flaskAmiibo[i]
    }

    override fun getItemViewType(position: Int): Int {
        return settings.amiiboView
    }

    fun getDuplicates(amiibo: Amiibo) : Int {
        return flaskAmiibo.count { it?.id == amiibo.id }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FlaskViewHolder {
        return when (VIEW.valueOf(viewType)) {
            VIEW.COMPACT -> CompactViewHolder(parent, settings, listener)
            VIEW.LARGE -> LargeViewHolder(parent, settings, listener)
            VIEW.IMAGE -> ImageViewHolder(parent, settings, listener)
            VIEW.SIMPLE -> SimpleViewHolder(parent, settings, listener)
        }
    }

    private fun setIsHighlighted(holder: FlaskViewHolder, position: Int) {
        val highlight = holder.itemView.findViewById<View>(R.id.highlight)
        if (mPrefs.flaskActiveSlot() == position) {
            highlight.setBackgroundResource(R.drawable.cardview_outline)
        } else {
            highlight.setBackgroundResource(0)
        }
    }

    override fun onBindViewHolder(holder: FlaskViewHolder, position: Int) {
//        setIsHighlighted(holder, position)
        holder.itemView.setOnClickListener {
            holder.listener?.onAmiiboClicked(holder.amiibo, holder.bindingAdapterPosition)
        }
        holder.imageAmiibo?.setOnClickListener {
            if (settings.amiiboView == VIEW.IMAGE.value)
                holder.listener?.onAmiiboClicked(holder.amiibo, holder.bindingAdapterPosition)
            else holder.listener?.onAmiiboImageClicked(holder.amiibo)
        }
        holder.bind(getItem(position))
    }

    abstract class FlaskViewHolder(
        itemView: View, private val settings: BrowserSettings, val listener: OnAmiiboClickListener?
    ) : RecyclerView.ViewHolder(itemView) {
        val txtError: TextView?
        val txtName: TextView?
        val txtTagId: TextView?
        val txtAmiiboSeries: TextView?
        val txtAmiiboType: TextView?
        val txtGameSeries: TextView?
        val txtPath: TextView?
        val txtUsage: TextView?
        var imageAmiibo: AppCompatImageView? = null
        var amiibo: Amiibo? = null

        var target: CustomTarget<Bitmap?> = object : CustomTarget<Bitmap?>() {
            override fun onLoadFailed(errorDrawable: Drawable?) {
                imageAmiibo?.setImageResource(R.drawable.ic_no_image_60)
            }

            override fun onLoadCleared(placeholder: Drawable?) {
                imageAmiibo?.setImageResource(R.drawable.ic_no_image_60)
            }

            override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap?>?) {
                imageAmiibo?.setImageBitmap(resource)
            }
        }

        init {
            txtError = itemView.findViewById(R.id.txtError)
            txtName = itemView.findViewById(R.id.txtName)
            txtTagId = itemView.findViewById(R.id.txtTagId)
            txtAmiiboSeries = itemView.findViewById(R.id.txtAmiiboSeries)
            txtAmiiboType = itemView.findViewById(R.id.txtAmiiboType)
            txtGameSeries = itemView.findViewById(R.id.txtGameSeries)
            txtPath = itemView.findViewById(R.id.txtPath)
            txtUsage = itemView.findViewById(R.id.txtUsage)
            imageAmiibo = itemView.findViewById(R.id.imageAmiibo)
        }

        fun bind(item: Amiibo?) {
            amiibo = item
            txtPath?.isGone = true
            txtUsage?.isGone = true
            var amiiboHexId = ""
            var amiiboSeries = ""
            var amiiboType = ""
            var gameSeries = ""
            var amiiboImageUrl: String? = null
            when (amiibo) {
                null -> {
                    setAmiiboInfoText(txtName, TagMo.appContext.getString(R.string.empty_tag))
                    txtError?.isGone = true
                    txtTagId?.isGone = true
                    txtAmiiboSeries?.isGone = true
                    txtAmiiboType?.isGone = true
                    txtGameSeries?.isGone = true
                    return
                }
                is FlaskTag -> {
                    setAmiiboInfoText(txtName, TagMo.appContext.getString(R.string.blank_tag))
                }
                else -> {
                    setAmiiboInfoText(txtName, amiibo?.name)
                    amiiboImageUrl = amiibo?.imageUrl
                }
            }
            imageAmiibo?.let {
                GlideApp.with(it).clear(it)
                if (amiiboImageUrl.isNullOrEmpty()) {
                    it.setImageResource(R.drawable.ic_no_image_60)
                } else {
                    GlideApp.with(it).asBitmap().load(amiiboImageUrl).into(target)
                }
            }
            if (settings.amiiboView != VIEW.IMAGE.value) {
                txtError?.isGone = true
                if (amiibo is FlaskTag) {
                    txtTagId?.isGone = true
                    txtAmiiboSeries?.isGone = true
                    txtAmiiboType?.isGone = true
                    txtGameSeries?.isGone = true
                } else {
                    amiibo?.let {
                        amiiboHexId = Amiibo.idToHex(it.id)
                        it.amiiboSeries?.let { series -> amiiboSeries = series.name }
                        it.amiiboType?.let { type -> amiiboType = type.name }
                        it.gameSeries?.let { series -> gameSeries = series.name }
                    }
                    txtTagId?.isVisible = true
                    txtAmiiboSeries?.isVisible = true
                    txtAmiiboType?.isVisible = true
                    txtGameSeries?.isVisible = true
                    setAmiiboInfoText(txtTagId, amiiboHexId)
                    setAmiiboInfoText(txtAmiiboSeries, amiiboSeries)
                    setAmiiboInfoText(txtAmiiboType, amiiboType)
                    setAmiiboInfoText(txtGameSeries, gameSeries)
                    if (hasSpoofData(amiiboHexId)) txtTagId?.isEnabled = false
                }
            }
        }

        fun setAmiiboInfoText(textView: TextView?, text: CharSequence?) {
            textView?.isVisible = true
             if (!text.isNullOrEmpty()) {
                textView?.text = text
                textView?.isEnabled = true
            } else {
                textView?.setText(R.string.unknown)
                textView?.isEnabled = false
            }
        }
    }

    internal class SimpleViewHolder(
        parent: ViewGroup, settings: BrowserSettings, listener: OnAmiiboClickListener?
    ) : FlaskViewHolder(
        LayoutInflater.from(parent.context).inflate(
            R.layout.amiibo_simple_card, parent, false
        ),
        settings, listener
    )

    internal class CompactViewHolder(
        parent: ViewGroup, settings: BrowserSettings, listener: OnAmiiboClickListener?
    ) : FlaskViewHolder(
        LayoutInflater.from(parent.context).inflate(
            R.layout.amiibo_compact_card, parent, false
        ),
        settings, listener
    )

    internal class LargeViewHolder(
        parent: ViewGroup, settings: BrowserSettings, listener: OnAmiiboClickListener?
    ) : FlaskViewHolder(
        LayoutInflater.from(parent.context).inflate(
            R.layout.amiibo_large_card, parent, false
        ),
        settings, listener
    )

    internal class ImageViewHolder(
        parent: ViewGroup, settings: BrowserSettings, listener: OnAmiiboClickListener?
    ) : FlaskViewHolder(
        LayoutInflater.from(parent.context).inflate(
            R.layout.amiibo_image_card, parent, false
        ),
        settings, listener
    )

    interface OnAmiiboClickListener {
        fun onAmiiboClicked(amiibo: Amiibo?, position: Int)
        fun onAmiiboImageClicked(amiibo: Amiibo?)
    }
}